01 02 MeetupPageDraft

ase444ka opened the pull request.

ase444ka committed:
Merge branch 'master' of github.com:js-tasks-ru/vue-20210119_istomina-asia into master

ase444ka committed:
01 02 MeetupPageDraft in progress
- made data receiving from API
- made hiding before having mounted
- made rendering div only if defined data

ase444ka committed:
01 02 MeetupPageDraft in progress 2
-Made images uploading from received data

ase444ka committed:
01 02 MeetupPageDraft in progress 3
-Made correct date rendering
-Made rendering of all data but agenda

ase444ka committed:
01 02 MeetupPageDraft complete
-made rendering agenda
-added comments
-little refactor
-added favicon

jsru commented:
Пожалуйста, не меняйте «системные» файлы.
          Задача не может считаться решенной, если были изменены файлы тестов или другие «системные» файлы.
          Следующие файлы были изменены:
```diff
- 01-basics/02-MeetupPageDraft/__tests__/MeetupPageDraft.test.js — добавлен
```

ase444ka commented:
> 
> 
> Пожалуйста, не меняйте «системные» файлы.
> Задача не может считаться решенной, если были изменены файлы тестов или другие «системные» файлы.
> Следующие файлы были изменены:
> 
> ```diff
> - 01-basics/02-MeetupPageDraft/__tests__/MeetupPageDraft.test.js — добавлен
> ```

Это как-то прямо очень странно! я точно не меняла/добавляла это файл о_О. вот фавикон добавила :) что теперь делать? :)

ase444ka commented:
`git diff master 89bbceb 01-basics/02-MeetupPageDraft/__tests__/MeetupPageDraft.test.js ` никакой разницы не выводит. не понимаю.

ase444ka committed:
01 02 MeetupPageDraft fix
-removed favicon

ase444ka commented:
Виноват, очевидно, фавикон. Поправила.

jsru commented:
Пожалуйста, не меняйте «системные» файлы.
          Задача не может считаться решенной, если были изменены файлы тестов или другие «системные» файлы.
          Следующие файлы были изменены:
```diff
- 01-basics/02-MeetupPageDraft/__tests__/MeetupPageDraft.test.js — добавлен
```

ase444ka committed:
Merge branch 'master' of github.com:js-tasks-ru/vue-20210119_istomina-asia into master

jsru commented:
Поздравляю, все задачи решены абсолютно верно
```diff
Meetup Page Draft
+ Задача добавлена, но требует ручной проверки
```

Добавляю преподавателя (@ShGKme) для код-ревью, посмотрим что он скажет :)

ShGKme commented on 01-basics/02-MeetupPageDraft/index.html:
```diff
@@ -7,32 +7,37 @@
+            <div class="meetup-agenda__item" v-for="item in agenda">
               <div class="meetup-agenda__item-col">
-                <img class="icon" alt="icon" src="/assets/icons/icon-cal-sm.svg" />
+                <img class="icon" alt="icon" :src="item.icon" v-if = "item.icon" />
```
Обособлять `=` пробелами во vue шаблонах не принято, также, как в HTML

ShGKme commented on 01-basics/02-MeetupPageDraft/index.html:
```diff
@@ -7,32 +7,37 @@
-                <h5 class="meetup-agenda__title">Название доклада</h5>
-                <p><span>Докладчик</span><span class="meetup-agenda__dot"></span><span class="meetup-agenda__lang">RU</span></p>
-                <p>Описание доклада</p>
+                <h5 class="meetup-agenda__title" v-if = "item.title" v-html = "item.title"></h5>
```
`v-html` выведет значение как html без какого-либо экранирования. Это потенциальный [XSS](https://ru.wikipedia.org/wiki/%D0%9C%D0%B5%D0%B6%D1%81%D0%B0%D0%B9%D1%82%D0%BE%D0%B2%D1%8B%D0%B9_%D1%81%D0%BA%D1%80%D0%B8%D0%BF%D1%82%D0%B8%D0%BD%D0%B3).

Любой пользователь может сделать свой митап с любым заголовком, в том числе использовать там теги или внедрять скрипты. С `v-html` получится серьёзная дыра в безопасности.

Правильнее использовать обычные выражения `{{ value }}`.

Если так было сделано из-за `<s>` в названии одного из докладов, то это оригинальное название: [https://vueconf.us/schedule/](https://vueconf.us/schedule/). `All you need is <s>love</s> Apollo Client`

Но наличие таких названий не должно приводить к XSS

ShGKme commented on 01-basics/02-MeetupPageDraft/script.js:
```diff
@@ -15,6 +19,22 @@ function getMeetupCoverLink(meetup) {
 }
 
+//получение даты вида 8 мая 2020
+function getLocaleDate(date) {
+  date = new Date(date);
+  const MONTHS = ['янв.', 'фев.', 'мар.', 'апр.', 'мая', 'июня', 'июля', 'авг.', 'сен.', 'нояб.', 'дек.'];
+  return `${date.getDate()} ${MONTHS[date.getMonth()]} ${date.getFullYear()} г.`
+}
```
Для локализованного вывода даты лучше использовать стандартный метод [`toLocalDateString()`](https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString).

На вебинаре мы использовали следующий код, который будет работать в соответствии с настройками локализации пользователя:
```javascript
new Date(meetup.date).toLocaleString(navigator.language, {
  year: 'numeric',
  month: 'long',
  day: 'numeric',
})
```

ShGKme commented on 01-basics/02-MeetupPageDraft/script.js:
```diff
@@ -48,19 +68,46 @@ export const app = new Vue({
-    //
+
+    meetup() {
+      return !this.rawMeetup ? {} :
```
Если данных митапа нет, лучше так и возвращать `null`, а не пустой объект `{}`. Пустой объект - это существующий объект, который логически приводится к `true`, но не имеет никаких данных, никаких свойств, даже тех, которые обязательно есть в объекте с данными митапа и не могут отсутствовать у митапа.
Это приводит к либо тяжело находящимся ошибкам, либо излишнему количеству проверок на каждое поле.

ShGKme commented on 01-basics/02-MeetupPageDraft/script.js:
```diff
@@ -48,19 +68,46 @@ export const app = new Vue({
+    },
+
+    agenda() {
+      return this.meetup.agenda && this.meetup.agenda.map(item => {
+          item.icon = `/assets/icons/icon-${agendaItemIcons[item.type]}.svg`;
+          item.title = item.title || agendaItemTitles[item.type]; //установка названия по умолчанию если нужно
+          item.isTalk = item.type === "talk"; //проверка выступление или нет - указывать ли спикера
+          item.period = `${item.startsAt} - ${item.endsAt}`;
+          return item;
+        })
```
Вычисляемые свойства - это такие свойства, для которых есть функция, вычисляющая их значение. Эта функция будет вызываться, когда требуется заново вычислить значение такого свойства, и в качестве значения свойства будет возвращаемое значение этой функции.

Вычисляемые свойства при своём вычислении **ни в коем случае не должны иметь побочных эффектов, не должны изменять состояние, не должны изменять другие свойства**.

При вычислении значения функция должна просто возвращать значение, вычисленное на основе текущего состояния (других свойств).

В данном случае правильно использовать метод `map`, который позволяет создать новый массив на основе старого, но при этом вы в нём изменяете каждое свойство, мутируете данные. Так делать ни в коем случае нельзя.

Требуется вернуть новый объект, а не изменять объекты в изначальном списке.

ShGKme reviewed (changes requested):
Требуется несколько исправлений:
1. Избавиться от проблемы с безопасностью, от XSS
2. Избавиться от побочных эффектов при вычислении вычисляемого свойства.
3. meetup должен быть `null`, а не `{}`, когда данных нет
4. Для локализации времени требуется использовать стандартные методы, учитывающие настройки локализации пользователя в браузере

ase444ka committed:
01 02 MeetupPageDraft fix
 -fixed problem with unsafety from xss
 -fixed problem with side effect at computed property
 -meetup now initializing by null - not {}
 -used standard date-method

ShGKme reviewed (approved):
Принято 👍
